### PR TITLE
Optimize type conversion functions. 

### PR DESCRIPTION
**Pull request by TristonianJones**
Unary type-conversion functions with literal arguments should be evaluated
eagerly when the `cel.EvalOptions(cel.OptOptimize)` flag is set. The conversion
will also ensure that if an error is encountered when evaluating these functions,
such as when constructing timestamps or durations from strings, the evaluation
will result in an error in the `env.Program` call.

This is not a perfect fix for #359, but it can be used to assist with correctness checks.

### CODE REVIEW

**Comment by glerchundi** [comment]
This is great! Correct me if i'm wrong, this implementation would contract the AST from a `timestamp('correct-ts-value')` into `*exprpb.Constant_TimestampValue`?

**Comment by lopezator** [comment]
Yahoo! 🎉

**Comment by TristonianJones** [comment]
@glerchundi well, not exactly. The code will convert from `timestamp(<rfc_3339_str>)` to a CEL `ref.Val`. The AST is left unmodified which is why this is an imperfect fix for your needs. You could use it as a work-around in the short term.

**Comment by glerchundi** [comment]
Ok, thanks anyway :D. Good job!

**Comment by TristonianJones** [comment]
@glerchundi I'll eventually get to supporting linting. This will at least tell you that the checked-AST has `timestamp(<str>)` and `duration(<str>)` calls with well-formed string inputs. It's not perfect, but it should meet your needs in the short-term. Just make sure to turn on optimization.

I have a few more tests that I need to add to catch malformed inputs at optimization time, but then I'll send this out for review.

**Comment by glerchundi** [comment]
I guess it does partially.

Currently we're parsing the expression and the walking the AST to build the SQL expression. Having an evaluation process where it executes the corresponding sanity check before proceeding sounds good but it's not enough as we need to continue walking the unmodified AST and taking care of the call expressions in it (like `timestamp`). 

Ideally I would expect `OptOptimize` to construct (i don't care the mechanism `newAst := PruneAst(details, oldAst)`) a cleaned up tree in order to avoid dealing with convert overloads on our side.

We're already doing it but makes the implementation more complicated that I would like it to be.

**Comment by glerchundi** [comment]
Does your decorator implementation in #363 helps converting/contracting `ref.Val` (with `timestamp(const)`) into another `ref.Val` (with timestamp_const_value). I must admit that i'm a bit lost there :D

**Comment by TristonianJones** [comment]
Hi @glerchundi, #363 is a decorator that participates in an AST walk done before evaluation. It's not really what you're looking for since it doesn't modify the AST; however, what I could do is expose a `ParserDecorator` which would allow you to observe, alter, or replace the `Expr` AST nodes at parse time. It would at least simplify your existing logic.

File: common/overloads/overloads.go [hunk 1]
```diff
@@ -271,3 +271,22 @@ const (
...
+	switch function {
+	case TypeConvertBool,
+		TypeConvertBytes,
+		TypeConvertDouble,
+		TypeConvertDuration,
+		TypeConvertInt,
+		TypeConvertString,
+		TypeConvertTimestamp,
+		TypeConvertType,
+		TypeConvertUint:
```
Comment (JimLarson) [inline common/overloads/overloads.go:288]: Consider whether TypeConvertDyn should also give a positive answer. If not, add a comment to explain the discrepancy versus the list of `TypeConvert*` constants above.

Note that the language spec lacks type conversion functions for `bool()` and `type()`, while it contains `list()` and `map()` conversions. Should the former be added to the spec?  Should the latter be added to the list here?  Looks like we implement a string-to-bool conversion that's not in the spec.

File: interpreter/interpreter_test.go [hunk 2]
```diff
@@ -278,6 +278,11 @@ var (
 			},
 			out: types.False,
 		},
+		{
+			name: "in_empty_list",
+			expr: `6 in []`,
+			out:  types.False,
+		},
```
Comment (JimLarson) [inline interpreter/interpreter_test.go]: This test seems unrelated to this PR - maybe move to another PR?

File: interpreter/interpreter_test.go [hunk 3]
```diff
@@ -1164,6 +1169,49 @@ func TestInterpreter_MissingIdentInSelect(t *testing.T) {
 	}
 }
 
+func TestInterpreter_TypeConversionOpt_Errors(t *testing.T) {
```
Comment (JimLarson) [inline interpreter/interpreter_test.go]: Add some positive tests that ensure that the optimization actually happens.

**Review by JimLarson — CHANGES_REQUESTED** [review]

File: common/overloads/overloads.go (hunk 1 above)
Comment (TristonianJones) [reply common/overloads/overloads.go:288]: It looks like `list()` and `map()` type conversions were never implemented. I'm actually having a hard time making sense of what they were intended to do. :/ They sort of look like identity functions:

```
list(T) -> list(T) // ???
```

With respect to the places where cel-go goes beyond the spec, the type conversions are usually pretty reasonable. I may have picked a strange name for `type(uint)` as `TypeConvertType`, but it's the standard type function: `type(uint) == type`, `type(null_type) == type`. Stuff like that.

I'll add support for `dyn()` as well.

File: interpreter/interpreter_test.go (hunk 2 above)
Comment (TristonianJones) [reply interpreter/interpreter_test.go]: Ah, sure thing. It must have gotten picked up when I was looking over coverage.

File: common/overloads/overloads.go (hunk 1 above)
Comment (TristonianJones) [reply common/overloads/overloads.go:288]: Per https://github.com/google/cel-spec/issues/123, I'll forego the `list()` and `map()` conversion functions, but will add `dyn()`. 

With respect to `bool()` I believe the identity function should be supported and I'm not opposed to supporting `string(bool("true")) == "true"` since this seems like reasonable round-tripping behavior even if its infrequently used.

**Comment by TristonianJones** [comment]
PTAL

**Comment by glerchundi** [comment]
That is a fantastic idea @TristonianJones!

I'll create an issue asking for exactly that and in case we found a free time slot & we see ourselves with the capacity to make it happen without messing the code we'll send a PR.

**Review by JimLarson — APPROVED** [review]
